Feature/disablepruning : Disable Jenkins Job Pruning via Annotation

waveywaves opened the pull request:
### Description

This feature allows the user to disable job pruning in jenkins if they annotate their buildconfig with `jenkins.openshift.io/disable-job-prune=true` as follows 
```
$ oc annotate bc nodejs-ex jenkins.openshift.io/disable-job-prune=true
```
Where nodejs-ex is being watched by the jenkins-sync-plugin.

wrt https://github.com/openshift/jenkins-sync-plugin/issues/312

### Acceptance Criteria

- [x] Add Annotation
- [x] Write the logic for disabling pruning
- [x] Update README

0 > 
![Screenshot from 2019-08-16 20-24-23](https://user-images.githubusercontent.com/11972233/63176672-fbc81580-c063-11e9-81ba-fb6fbbb7d53f.png)

![Screenshot from 2019-08-16 20-26-41](https://user-images.githubusercontent.com/11972233/63176784-32059500-c064-11e9-9812-6e3cca4d540b.png)

1 > Annotating the BuildConfig :-
`oc annotate bc nodejs-ex jenkins.openshift.io/disable-job-prune=true`

2 > Deleting Build: 
`oc delete build nodejs-ex-27`
```
INFO: Reconciling job runs and builds
Aug 16, 2019 2:56:13 PM io.fabric8.jenkins.openshiftsync.BuildWatcher reconcileRunsAndBuilds
INFO: Checking job org.jenkinsci.plugins.workflow.job.WorkflowJob@30553b6a[jenkins-sync-test/jenkins-sync-test-nodejs-ex] runs for BuildConfig jenkins-sync-test/nodejs-ex                                        
Aug 16, 2019 2:56:13 PM io.fabric8.jenkins.openshiftsync.JobProcessor createOrUpdateJob
INFO: Updated job jenkins-sync-test-nodejs-ex from BuildConfig NamespaceName{jenkins-sync-test:nodejs-ex} with revision: 413390                                                                                   
Aug 16, 2019 2:58:04 PM io.fabric8.jenkins.openshiftsync.JenkinsUtils deleteRunIfPruneDisabled
INFO: Prune disabled on Jenkins Job jenkins-sync-test/jenkins-sync-test-nodejs-ex
``` 
![Screenshot from 2019-08-16 20-41-51](https://user-images.githubusercontent.com/11972233/63177828-5c585200-c066-11e9-982e-9ccf4a922050.png)

![Screenshot from 2019-08-16 20-43-06](https://user-images.githubusercontent.com/11972233/63177881-78f48a00-c066-11e9-8332-5ddcf1b9a1a5.png)

waveywaves commented:
/assign @akram 
/assign @gabemontero

akram commented:
/hold

gabemontero commented on src/main/java/io/fabric8/jenkins/openshiftsync/BuildWatcher.java:
```diff
@@ -491,6 +493,8 @@ private static void reconcileRunsAndBuilds() {
       BuildList buildList = getAuthenticatedOpenShiftClient().builds()
         .inNamespace(property.getNamespace()).withLabel("buildconfig=" + property.getName()).list();
 
+      BuildConfig buildConfig = property.getBuildConfig();
```
in line with @akram 's comment below, you should be getting a fresh version of the buildconfig vs. pulling the copy stored in the property object

waveywaves commented:
I have updated the functionality. We are now able to check if the display name contains the `DISABLE_PRUNE_PREFIX` and then disable pruning for the particular WorkflowRun. Also,t his prefix is added to the runs when deletion is triggered on the builds. 

![Screenshot from 2019-08-16 18-58-40](https://user-images.githubusercontent.com/11972233/63173012-97558800-c05c-11e9-9860-722fcc531d9a.png)

I have changed the prefix to disabledPrune right now.

waveywaves commented:
/retest

waveywaves commented:
Before adding the annotation : 

![Screenshot from 2019-08-16 20-24-23](https://user-images.githubusercontent.com/11972233/63176672-fbc81580-c063-11e9-81ba-fb6fbbb7d53f.png)

Annotating the BuildConfig:
`oc annotate bc nodejs-ex jenkins.openshift.io/disable-job-prune=true`

waveywaves commented:
Builds under a particular BC will have their display name prefixed with pruneDisabled and they won't be deleted even if annotation is not present given that they use the version of sync plugin this feature is released in.

gabemontero commented on src/main/java/io/fabric8/jenkins/openshiftsync/BuildWatcher.java:
```diff
@@ -482,29 +485,26 @@ private static void reconcileRunsAndBuilds() {
         }
         if (!found) {
-          deleteRun(run);
+            deleteRunIfPruneDisabled(run, buildConfig);
```
should this be `deleteRunIfPruneEnabled`

gabemontero replied:
and I mean from a naming perspective ... I see from the compile error at https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_jenkins-sync-plugin/325/pull-ci-openshift-jenkins-sync-plugin-master-e2e-aws-jenkins/72#0:build-log.txt%3A3929 you at least considered that at some point

waveywaves replied:
I changed everything to Enabled. Got a little confused iwth the name while writing this. Thank you Gabe :)

waveywaves commented:
@gabemontero Updated README

akram commented on src/main/java/io/fabric8/jenkins/openshiftsync/JenkinsUtils.java:
```diff
@@ -545,9 +555,11 @@ public static void deleteRun(WorkflowRun run) {
 
+
   public static void deleteRun(WorkflowJob job, Build build) {
       WorkflowRun run = getRun(job, build);
```
probably, we should set the deleteRunIfPruneEnabled method call the deleteRun method, and not the opposite.
Let's discuss it.

waveywaves replied:
We would need the BuildConfigProjectProperty in this case which I wasn't able to get using the Run. Hence the wrapper.

akram commented on src/main/java/io/fabric8/jenkins/openshiftsync/JobProcessor.java:
```diff
@@ -109,14 +111,38 @@ private void populateNamespaceFolder(Jenkins activeInstance, ItemGroup parent, S
 	}
 
+	private BuildConfigProjectProperty updateUID(BuildConfigProjectProperty buildConfigProjectProperty, BuildConfigProjectProperty newProperty){
+    String UID = buildConfigProjectProperty.getUid();
```
better to name it uid instead of UID.

akram commented on src/main/java/io/fabric8/jenkins/openshiftsync/JobProcessor.java:
```diff
@@ -109,14 +111,38 @@ private void populateNamespaceFolder(Jenkins activeInstance, ItemGroup parent, S
+	private BuildConfigProjectProperty updateUID(BuildConfigProjectProperty buildConfigProjectProperty, BuildConfigProjectProperty newProperty){
+    String UID = buildConfigProjectProperty.getUid();
+    String newUID = newProperty.getUid();
+    if (!UID.contains(DISABLE_PRUNE_PREFIX)){
```
Here, we should be able to nest the 2 conditions:

```
if ( !uid.startsWith( PRUNING_DISABLED_PREFIX) ) {
  if ( !isJobPruningDisabled(buildConfigProjectProperty) ) {
    doSomething();
  } else {
    doOtherThing();
  }
}
```

akram commented on src/main/java/io/fabric8/jenkins/openshiftsync/JobProcessor.java:
```diff
@@ -109,14 +111,38 @@ private void populateNamespaceFolder(Jenkins activeInstance, ItemGroup parent, S
+      logger.info("Prune disabled on Jenkins Job " + jenkinsJobFullName(buildConfig));
+    }
+
+    String oldUID = UID.replace(DISABLE_PRUNE_PREFIX,"");
```
We will specify this part more precisely, but, looks as a good start

akram commented on src/main/java/io/fabric8/jenkins/openshiftsync/JobProcessor.java:
```diff
@@ -109,14 +111,38 @@ private void populateNamespaceFolder(Jenkins activeInstance, ItemGroup parent, S
+    System.out.println("-------------- Comparing names "+jenkinsJobFullName(buildConfig)+"------------ "+oldName);
+    System.out.println("-------------- Comparing UID -------"+UID+"--------"+oldUID+"--------"+newUID);
+
+    if (UID.contains(DISABLE_PRUNE_PREFIX) && !oldUID.equals(newUID) && oldName.equals(jenkinsJobFullName(buildConfig))){
```
I am not sure that we need to check that the prefix is present here: In case of the a first synchronisation, we should be able to detect at the "first pass" that the name is the same but uid differences, and the annotation is present, so, we override the uid ?


and this also applies with a second sync, if the prefix is here, we should also try to sync and update the uid if relevant. 
Am I correct?

akram commented on src/main/java/io/fabric8/jenkins/openshiftsync/JobProcessor.java:
```diff
@@ -160,26 +186,32 @@ private void populateNamespaceFolder(Jenkins activeInstance, ItemGroup parent, S
-			buildConfigProjectProperty.setNamespace(newProperty.getNamespace());
+
+      updateUID(buildConfigProjectProperty, newProperty);
+      buildConfigProjectProperty.setNamespace(newProperty.getNamespace());
```
Maybe we need a test to surround this one. Also, because the namespace name is contained in the directory name inside of the Jenkins volume /var/lib/jenkins/<namespacename>-<buildconfigname>

Maybe, in some cases, we will need to rename the directory.

waveywaves commented:
/test e2e-aws-jenkins
